### PR TITLE
[select] Fire onClick during drag-to-select

### PR DESCRIPTION
**Pull request by obeattie**
- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/base-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

The drag-to-select behavior which was originally introduced in #541 (mousedown on trigger → drag to item → mouseup to select) bypasses the item's `onClick` handler since the browser doesn't fire a `click` event when mousedown and mouseup occur on different elements. This is inconsistent with the other ways to select an item: both regular clicks and keyboard selection both invoke `onClick` already.

This PR calls `onClick` from `onMouseUp` before committing the selection, allowing `preventBaseUIHandler()` to cancel the selection.

Includes tests. (Sorry about #3968 - I closed it by accident.)

### CODE REVIEW

File: packages/react/src/select/item/SelectItem.tsx [hunk 1]
```diff
@@ -229,7 +236,12 @@ export const SelectItem = React.memo(
           return;
         }
 
-        commitSelection(event.nativeEvent);
+        makeEventPreventable(event);
+        componentProps.onClick?.(event);
+
+        if (!event.baseUIHandlerPrevented) {
+          commitSelection(event.nativeEvent);
+        }
```
Comment (atomiks) [inline packages/react/src/select/item/SelectItem.tsx]: In `MenuItem`, we call `itemRef.current.click()` instead, I wonder if that's better for consistency

**Comment by obeattie** [comment]
@atomiks Thanks for taking a look. Agree that doing `itemRef.current.click()` is the right move here. I'd actually tried that initially, but it exposed another bug that I've now gone ahead and fixed here.

Specifically, when a `Select` is nested inside a `Popover`, drag-to-select when implemented like this was unexpectedly dismissing the popover. What was happening is that the “started inside” drag state could be consumed too early. The fix is to keep that state through click capture, check whether the click target is inside _first_, and only use the one-shot suppression on the first real outside click.

This also shows up outside `Select`. For example, see this (ugly) Stackblitz that exhibits it with a `Menu` inside a `Popover`: https://stackblitz.com/edit/gzrjdbd3?file=src%2FApp.tsx

File: packages/react/src/select/item/SelectItem.tsx [hunk 2]
```diff
@@ -208,11 +214,12 @@ export const SelectItem = React.memo(
         pointerTypeRef.current = event.pointerType;
         didPointerDownRef.current = true;
       },
-      onMouseUp(event) {
+      onMouseUp(event: BaseUIEvent<React.MouseEvent<HTMLDivElement>>) {
```
Comment (atomiks) [inline packages/react/src/select/item/SelectItem.tsx]:
```suggestion
      onMouseUp(event) {
```

**Comment by atomiks** [comment]
> but it exposed another bug that I've now gone ahead and fixed here.

Oh nice, thanks for the fix & explanation

**Comment by atomiks** [comment]
Couple lint issues:

> [ci/circleci: Linting](https://circleci.com/gh/mui/base-ui/222107) — Your tests failed on CircleCI

File: packages/react/src/floating-ui-react/hooks/useDismiss.test.tsx [hunk 3]
```diff
@@ -868,6 +868,18 @@ describe.skipIf(!isJSDOM)('useDismiss', () => {
       fireEvent.click(document.body);
       expect(screen.queryByRole('tooltip')).not.toBeInTheDocument();
     });
+
+    test('inside click does not suppress the next outside click when no drag occurred', async () => {
```
Comment (atomiks) [inline packages/react/src/floating-ui-react/hooks/useDismiss.test.tsx]: Codex review found a logic regression here:

> a normal inside click in intentional mode can leave endedOrStartedInsideRef set,
so the next outside click is incorrectly ignored once (even without a drag).

I can't reproduce it this on the Popover demo, though. The tests are only failing in JSDOM but not Chromium.

Comment (atomiks) [reply packages/react/src/floating-ui-react/hooks/useDismiss.test.tsx]: @obeattie this seems like an edge case, but theoretically possible to trigger an issue

In-browser popover usage closes because real browser clicks fire the other mouse events (mousedown -> mouseup -> click). And in useDismiss, non-click events clear `endedOrStartedInsideRef`, so by the time outside click runs, it closes normally. 

So it's an issue only if `.click()` runs without the mouse events (tests, assistive tech, programmatic clicks?)

Comment (obeattie) [reply packages/react/src/floating-ui-react/hooks/useDismiss.test.tsx]: Thanks for the review and good catch. The only sane way I could think of to resolve this was to change the suppression logic to better tie the suppression to actual drag sequences.

Now, two flags are used for this:

* `pressStartedInsideRef`: Tracks whether there's an active press that began inside. Always cleared in `pointerup`/`mouseup`
* `suppressNextOutsideClickRef`: Tracks whether we want to ignore an outside `click` event – set on `pointerup`/`mouseup` (which fire before `click` and won't fire in the synthetic cases like the one we're now firing for drag-to-select).

Hopefully this works better!

**Comment by atomiks** [comment]
Simplified the `useDismiss` logic further to avoid a (minzip) bundle size increase, some existing logic wasn't necessary

**Review by atomiks — APPROVED** [review]

**Comment by atomiks** [comment]
@greptileai

File: packages/react/src/select/item/SelectItem.tsx [hunk 4]
```diff
@@ -229,7 +230,7 @@ export const SelectItem = React.memo(
           return;
         }
 
-        commitSelection(event.nativeEvent);
+        itemRef.current?.click();
       },
```
[the remark this answers is not in the document]
Comment (atomiks) [reply packages/react/src/select/item/SelectItem.tsx:237]: Appears invalid

**Comment by obeattie** [comment]
Anything more I can do to help here?

**Comment by atomiks** [comment]
@obeattie nope, sorry, I edited your PR and then found various issues caused by my edits, and spent awhile refining it the other day. I _think_ its current state is good but I left it off the last release because it changes `useDismiss` significantly and wanted to manually test it more thoroughly

**Comment by obeattie** [comment]
No worries - thank you for your effort!
